### ISSUE TITLE
Buildtime compile

### ISSUE THREAD

**Issue by qm3ster** [glathoud/transfun#6]
Is there (an easy way to implement|plans to include|another project) that does this transform from functional methods to one loop as a build step?

**Comment by glathoud** [comment]
Thanks for the idea and apologies - somehow I completely missed that issue!! Sorry!

Now, on the top of my head: 

## TL;DR

Yes, easy, have a look at:
```js
var ff = tfun.map( '.p' ).filter( '!=null' ).reduce( '+' );

console.log( ff.getFunSpec().funSpec.varNameArr );
console.log( ff.getBodyCode() )
console.log( ff.getNExternal() ) // 0
```
and:
```js
var ff=tfun.map( x => x.p ).filter( x => x!=null ).reduce( (x,y)=>x+y )

console.log( ff.getFunSpec().funSpec.varNameArr );
console.log( ff.getBodyCode() )
console.log( ff.getNExternal() ) // 3
```

## Codestrings

At least in the "codestring" use case this should be relatively easy:
```js
tfun.map( '.p' ).filter( '!=null' ).reduce( '+' )
```
the build could detect such sequences and replace them with a generated code. As it is today, transfun already gives access to the necessary bits and parts:
```js
var ff = tfun.map( '.p' ).filter( '!=null' ).reduce( '+' );

console.log( ff.getFunSpec().funSpec.varNameArr );
// ["current"]
// (the list of input variables for that function, in that case only one variable)

console.log( ff.getBodyCode() )
`    var n = current.length;
    var out = null;
    var redinit = false;
    for (var k = 0; k < n; k++ ) {
        var v = current[ k ];
        
        /* -- map( `.p` ) -- */
        v = (v.p);
        
        /* -- filter( `!=null` ) -- */
        if ((v!=null)) {
            
            /* -- reduce( `+` ) -- */
            if (redinit) {
                out = (out+v);
            } else {
                out = v;
                redinit = true;
            }
        }
    }
    return out;`
```

## Function expressions

In that case:
```js
ff=tfun.map( x => x.p ).filter( x => x!=null ).reduce( (x,y)=>x+y )
```
transfun generates code that is bound to these external functions:
```js
ff.getFunSpec().funSpec.varNameArr
`["__extern_0__", "__extern_1__", "__extern_2__", "current"]`

ff.getBodyCode()
`    var n = current.length;
    var out = null;
    var redinit = false;
    for (var k = 0; k < n; k++ ) {
        var v = current[ k ];
        
        /* -- map( `__extern_0__` ) -- */
        v = __extern_0__.call(this,v,k);
        
        /* -- filter( `__extern_1__` ) -- */
        if (__extern_1__.call(this,v,k)) {
            
            /* -- reduce( `__extern_2__` ) -- */
            if (redinit) {
                out = __extern_2__.call(this,out,v,k);
            } else {
                out = v;
                redinit = true;
            }
        }
    }
    return out;`
```

So the build system would have to generate something like:
```js
var ff=(function (__extern_0__,__extern_1__,__extern_2__){
return function (current)
{
 var n = current.length;
    var out = null;
    var redinit = false;
    for (var k = 0; k < n; k++ ) {
      var v = current[ k ];
        
        /* -- map( `__extern_0__` ) -- */
        v = __extern_0__.call(this,v,k);
        
        /* -- filter( `__extern_1__` ) -- */
        if (__extern_1__.call(this,v,k)) {
            
            /* -- reduce( `__extern_2__` ) -- */
            if (redinit) {
                out = __extern_2__.call(this,out,v,k);
            } else {
                out = v;
                redinit = true;
            }
        }
    }
    return out;
};
})(x => x.p,x => x!=null,(x,y)=>x+y);
```

The code should run at least as fast as in the runtime variant. Whether this would bring an additional speedup in current JS engines, I don't know.

## Function declarations & function variables

Pretty much the same as for function expressions:
```js
var ff=tfun.map( f1 ).filter( f2 ).reduce( f3 )
```
would be replaced with:
```js
var ff=(function (__extern_0__,__extern_1__,__extern_2__){
   return function(current){ ... };
})(f1,f2,f3);
```

## Mixing

All three variants can be mixed. I don't see any problem here. My overall impression is that there shouldn't be any big technical issue integrating "transfun buildtime compilation" into a build system.

## Syntactic sugar: tval

The shortcut  `tval( v, tfun.[...] ) == (tfun.[...])( v )` would also need to be replaced with `(function (..) { ... })( v )`, but at that point that should be easy to do.

## Discussion

As with everything technical, I guess the question remains: *why* doing "transfun buildtime compilation"? In which context?

In the web context, the costs and benefits (esp. in terms of KB) have to be weighted against shipping a minified version of https://github.com/glathoud/transfun/blob/master/transfun.js inside the build.

Note that I am only talking about KB, because the actual CPU costs of running the transfun compilation itself are negligible, so from the CPU perspective it does not matter whether it runs at buildtime or at runtime. Only the memory costs and memory benefits are meaningful here.

If bandwidth is not an issue (e.g. node backend), or if you gain much more than you lose (*), nothing speaks against buildtime compilation.

(*) == using transfun only in a few places, in which case shipping transfun.js itself would be overkill ; but then again why using transfun only sparingly? I guess the pragmatic answer in the case where bandwidth matters is to let the build try both variants (a) leaving code as is and shipping transfun.js with the code and (b) buildtime compilation, and pick the smallest.

## Personally

At the moment I am not planning to do this, probably because I am addicted to runtime JavaScript code generation - pure dope for me, see e.g. #2 - BTW I have already been using transfun for fast prototyping of number crunching applications, writing fast code that runs fast (D spirit...) but I digress.

However, if anyone wants to do this, I'd be happy to provide the necessary support, if any. Again, I don't think that this should be very hard.

**Comment by glathoud** [comment]
Note that transfun also gives access to the number of externals:
```js
var ff_0 = tfun.map( '.p' ).filter( '!=null' ).reduce( '+' );
console.log(ff_0.getNExternal())  // 0

var ff_3=tfun.map( x => x.p ).filter( x => x!=null ).reduce( (x,y)=>x+y )
console.log(ff_3.getNExternal())  // 3
```

**Comment by glathoud** [comment]
Been wondering why I did not notice this last year... Maybe because I wasn't assigned to the issue ?! No idea.

**Comment by glathoud** [comment]
the most I can do here is build a node package #8

**Comment by glathoud** [comment]
@qm3ster FYI first steps going in the direction of buildtime compilation: the [somewhat long response above](https://github.com/glathoud/transfun/issues/6#issuecomment-399325451) and a [Node.js package](https://www.npmjs.com/package/transfun).
